### PR TITLE
Add option to log payloads on error response from the trace agent

### PR DESCRIPTION
**Pull request by brettlangdon**
Adding new environment variable `_DD_TRACE_WRITER_LOG_ERROR_PAYLOADS`
which when enabled will update the existing "failed to send traces to Datadog agent" log message
to include the payload.

The purpose of this is to debug any trace agent decoding issues.
With this we will be able to validate that the payloads
generated are valid and complete messages as expected.

The resulting log will look like:

```
ERROR:ddtrace.internal.writer:failed to send traces to Datadog Agent at http://localhost:8126: HTTP error status 400, reason Bad Request, payload <hex_encoded_mspack_binary_data>
```

You can then decode via:

```python
import binascii
import msgpack

payload = msgpack.unpackb(binascii.unhexlify("<hex_encoded_mspack_binary_data>"))
```

## Commit Message
{{title}}

Adding new environment variable `_DD_TRACE_WRITER_LOG_ERROR_PAYLOADS`
which when enabled will update the existing "failed to send traces to Datadog agent" log message
to include the payload.

### CODE REVIEW

**Comment by P403n1x87** [comment]
> You can then decode via:

If it helps, I normally pull in this bit of code when debugging `msgpack` related issues as it makes it slightly more obvious where the problems might be

~~~ python
def dump(data):
    W = 16
    F = "%%-%ds  |  %%s" % (W * 3 - 1)
    try:
        printable = str.isprintable
    except AttributeError:
        import string

        printable = lambda _: _ in string.printable and ord(_) > 31
    chunks = [data[i : i + W] for i in range(0, len(data), W)]
    print("")
    for c in chunks:
        _hex = " ".join(["%02x" % _ for _ in c])
        text = "".join(
            [chr(_) if printable(chr(_)) else "\033[30;1m.\033[30;0m" for _ in c]
        )
        print(F % (_hex, text))
    print("")
~~~

It gives you a side-by-side view of the hex codes and printable character representation, like most hex editors do.

**Comment by brettlangdon** [comment]
oh, cool, nice!

File: ddtrace/internal/writer.py [hunk 1]
```diff
@@ -141,7 +143,7 @@ def get_json(self):
                 )
                 return
 
-            return loads(body)
+            return json.loads(body)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:146]: nit: this json change looks to be unnecessary

File: ddtrace/internal/writer.py [hunk 2]
```diff
@@ -300,6 +302,7 @@ def __init__(
             stop=tenacity.stop_after_attempt(self.RETRY_ATTEMPTS),
             retry=tenacity.retry_if_exception_type((compat.httplib.HTTPException, OSError, IOError)),
         )
+        self._log_payloads = asbool(os.environ.get("_DD_TRACE_WRITER_LOG_ERROR_PAYLOADS", False))
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:305]:
```suggestion
        self._log_error_payloads = asbool(os.environ.get("_DD_TRACE_WRITER_LOG_ERROR_PAYLOADS", False))
```

**Review by Kyle-Verhoog — APPROVED** [review]

File: ddtrace/internal/writer.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/internal/writer.py:146]: oh, good catch. I did have a json usage in the original version. I'll adjust

File: ddtrace/internal/writer.py (hunk 2 above)
Comment (brettlangdon) [reply ddtrace/internal/writer.py:305]: good catch

File: tests/integration/test_integration.py [hunk 3]
```diff
@@ -393,6 +393,74 @@ def encode_traces(self, traces):
...
+            400,
+            "Bad Request",
+            "6261645f7061796c6f6164",
+        )
+    ]
+    log.error.assert_has_calls(calls)
+
+
+@pytest.mark.skipif(AGENT_VERSION == "testagent", reason="FIXME: Test agent response is different.")
+def test_bad_payload_log_payload_non_bytes(monkeypatch):
```
Comment (Kyle-Verhoog) [inline tests/integration/test_integration.py:431]: We could parametrize these but I don't mind personally

**Review by Kyle-Verhoog — APPROVED** [review]

File: ddtrace/internal/writer.py [hunk 4]
```diff
@@ -407,12 +410,22 @@ def _send_payload(self, payload, count):
...
                 self.agent_url,
                 response.status,
                 response.reason,
             )
+            # Append the payload if requested
+            if self._log_error_payloads:
+                msg += ", payload %s"
+                # If the payload is bytes then hex encode the value before logging
+                if isinstance(payload, six.binary_type):
+                    log_args += (binascii.hexlify(payload).decode(),)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:424]: this is probably going to be quite expensive with large payloads but i guess this is to be expected

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**